cpp/src_prims/distance/cosine.cuh

```diff
@@ -16,15 +16,167 @@
+ * @param[in]       n number of columns of B and C/D
+ * @param[in]       k number of cols of A and rows of B
+ * @param[output]   pD output matrix
+ * @param core_op   the core lambda
+ * @param epilog_op the epilogue lambda
+ * @param fin_op    the final gemm epilogue lambda
+ */
+template <typename DataT, typename AccT, typename OutT, typename IdxT,
+          typename Policy, typename CoreLambda, typename EpilogueLambda,
+          typename FinalLambda>
+__global__ __launch_bounds__(Policy::Nthreads, 2) void cosineKernel(
+  const DataT *x, const DataT *y, const DataT *_xn, const DataT *_yn, IdxT m,
+  IdxT n, IdxT k, OutT *dOutput, CoreLambda core_op, EpilogueLambda epilog_op,
+  FinalLambda fin_op) {
+  extern __shared__ char smem[];
+
+  PairwiseDistances<DataT, AccT, OutT, IdxT, false, Policy, CoreLambda,
+                    EpilogueLambda, FinalLambda>
+    obj(x, y, m, n, k, _xn, _yn, dOutput, smem, core_op, epilog_op, fin_op);
+  obj.run();
+}
+
+/**
+ * @brief the cosine distance matrix calculation implementer
+ *  It computes the following equation: 
+ *    C = 1 - op(A * B / sqrt(A^2) * sqrt(B^2)))
+ * @tparam DataT input data-type (for A and B matrices)
+ * @tparam AccT   accumulation data-type
+ * @tparam OutT   output data-type (for C and D matrices)
+ * @tparam IdxT   index data-type
+ * @tparam Veclen number of k-elements loaded by each thread for every LDG call
+ *                it makes. check contractions.cuh for details.
+ * @tparam FinalLambda the final lambda called on final distance value
+ * @param[in]     x input matrix
+ * @param[in]     y input matrix
+ * @param[in]     xn row norms of input matrix A.
+ * @param[in]     yn row norms of input matrix B.
+ * @param[in]     m number of rows of A and C/D
+ * @param[in]     n number of columns of B and C/D
+ * @param[in]     k number of cols of A and rows of B
+ * @param[output] pD output matrix
+ * @param fin_op  the final gemm epilogue lambda
+*  @param stream  cuda stream to launch cuda operations.
+ */
+template <typename DataT, typename AccT, typename OutT, typename IdxT,
+          int VecLen, typename FinalLambda>
+void cosineImpl(const DataT *x, const DataT *y, const DataT *xn,
+                const DataT *yn, IdxT m, IdxT n, IdxT k, OutT *dOutput,
+                FinalLambda fin_op, cudaStream_t stream) {
+  typedef typename raft::linalg::Policy4x4<DataT, VecLen>::Policy Policy;
+  dim3 grid(raft::ceildiv<int>(m, Policy::Mblk),
+            raft::ceildiv<int>(n, Policy::Nblk));
+  dim3 blk(Policy::Nthreads);
+
+  // Accumulation operation lambda
+  auto core_lambda = [] __device__(AccT & acc, DataT & x, DataT & y) {
+    acc += x * y;
+  };
+
+  // epilogue operation lambda for final value calculation
+  auto epilog_lambda = [xn, yn, m, n, k] __device__(
+                         AccT acc[Policy::AccRowsPerTh][Policy::AccColsPerTh],
+                         DataT * sxNorm, DataT * syNorm) {
+    __syncthreads();  // so that we can safely reuse smem
```

Comment:
As I mentioned above, I think we can avoid having to rely on users to do syncthreads by explicitly issuing it after `loop()` call in the pairwise distances class.